File: 720-longest-word-in-dictionary/longest-word-in-dictionary.cpp
class Solution {
public:
    string longestWord(vector<string>& words) {
        int n=words.size();

        sort(words.begin(),words.end());

        unordered_set<string>st;
        string ans="";
        for(int i=0;i<n;i++){
            string s=words[i];
            s.pop_back();

            if(words[i].length()==1 || st.count(s)){
                st.insert(words[i]);

                if(s.length()+1>ans.length()){
                    ans=words[i];
                }
            }
        }
        return ans;
    }
};

File: 720-longest-word-in-dictionary/longest-word-in-dictionary.java
class Solution {
    public String longestWord(String[] words) {
        int n=words.length;

        HashSet<String>st=new HashSet<>();

        Arrays.sort(words);
        String ans="";
        for(int i=0;i<n;i++){
            String s=words[i];

            s=s.substring(0,words[i].length()-1);

            if(words[i].length()==1 || st.contains(s)){
                st.add(words[i]);

                if(s.length()+1>ans.length()){
                    ans=words[i];
                }
            }
        }
        return ans;
    }
}